proposals/20210707-xgboost_evaluator.md

```diff
@@ -0,0 +1,119 @@
+
+# Project Proposal
+
+# Project Description
+Add support for evaluating XGBoost model in the standard component Evaluator. 
+Add an example pipeline that trains, evaluates and pushes an XGBoost model to CAIP.
+
+## Project Category
+
+Component + Example
+
+## Project Use-Case(s)
+
+## Project Implementation
+
+To make the Evaluator works with XGBoost models, we can customize the Evaluator by providing a Python module with:
+* `custom_eval_shared_model()` to load model artifacts that are not standard TF models
+* `custom_extractors()` to inject a custom prediction extractor. Similar to (tfma.extractors.PredictExtractor)[https://www.tensorflow.org/tfx/model_analysis/api_docs/python/tfma/extractors/PredictExtractor], this extractor uses Beam PTransform to load and extract predictions.
+
+### From Trainer to Evaluator: save and load XGBoost model
+
+Option 1: working with XGBoost library directly
+
+The XGBoost library provides a few different ways to save a model (an xgb.Booster or xgb.sklearn.XGBModel object). Backward compatibility is guaranteed in most cases. Currently, the 2 main supported formats are:
+* XGBoost internal binary format. Note that Auxiliary attributes of the Python Booster object (such as feature_names) will not be loaded when using binary format.
+* JSON: newer format aiming to replace the binary format
+
+For maximum compatibility, we want the Trainer component to output both formats at the expected output directory, and can provide a helper function, which takes in a Booster object then writes `model.bin` and `model.json` to the expected directory. 
+The Evaluator uses the latest version of the xgboost library to read `model.json` - this will be implemented in UDF `custom_eval_shared_model()`. This way, we can expect the loaded model object to have most necessary information retained.
+
+Option 2: using [sklearn Pipeline](https://scikit-learn.org/stable/modules/compose.html)
+
+```
+from sklearn.preprocessing import StandardScaler
+from sklearn.pipeline import Pipeline as SkPipeline
+
+classifier = xgb.XGBClassifier(**params)
+model = SkPipeline([
+    ('scaler', StandardScaler()),
+    ('classifier', classifier),
+])
+model.fit(x_train, y_train)
+
+# you can choose to save just the XGBClassifier:
+model.steps[1][1].save_model(...)
+```
+However, it’s more likely that you’ll need the whole sklearn Pipeline in downstream evaluation. There are 2 methods to save and load sklearn Pipeline object:
+
+Using joblib:
+```
+import joblib
+joblib.dump(pipeline, 'model.joblib')
+```
+Note that CAIP asks users to use `sklearn.externals.joblib` rather than the bare `joblib`, but newer versions of sklearn have deprecated `skearn.externals`.
+
+Using pickle:
+```
+import pickle
+with open('model.pkl', 'wb') as model_file:
+    pickle.dump(pipeline, model_file)
+```
+The main downside of working with sklearn Pipeline is potentially losing portability, this will be discussed further in the summary section.
```

Comment:
@TheMichaelHu to continue conversation from https://groups.google.com/a/tensorflow.org/g/sig-tfx-addons/c/Bx6zvqTVOU0

I agree w using Option 1 here (we can remove Option 2)